src/conv/invokers/mlir_impl_gemm.cpp

```diff
@@ -332,36 +360,84 @@ InvokerFactory MakeMlirWrWInvokerFactory(const ConvolutionContext& ctx)
                            weights_strides,
                            out_dims,
                            out_strides);
-    MlirConvArgs args =
-        MakeMlirConvArgs(in_dims, in_strides, weights_dims, weights_strides, out_dims, out_strides);
+    // Explicitly assume we will use a workspace, and prepare the
+    // StridedMemRef5D field inside MlirConvArgs structure.
+    // The reasons are:
+    // a) The logic to determine whether a config requires a workspace is
+    //    non-trivial inside MLIR kernel generator. We need to minimize the
+    //    number of call sites.
+    // b) The solver, not the invoker factor, will be responsible asking
+    //    MLIR kernel generator to compute the size of workspace needed.
+    // c) In the nested lambda returned below, the actual size of workspace
+    //    and the pointer will be supplied by the solver.
+    // d) In case a workspace is not needed, the pointer of a workspace will
+    //    not be passed to MLIR-generated kernel.
+    MlirConvArgs args = MakeMlirConvArgs(in_dims,
+                                         in_strides,
+                                         weights_dims,
+                                         weights_strides,
+                                         out_dims,
+                                         out_strides,
+                                         /*populateWorkspaceArg=*/true);
 
     return [=](const std::vector<Kernel>& kernels) mutable {
         return [=](const Handle& handle, const AnyInvokeParams& primitive_parameters) mutable {
             const auto& wrw_invoke_params = primitive_parameters.CastTo<conv::WrWInvokeParams>();
             const auto& tensors           = wrw_invoke_params.tensors;
+            const auto& workspaceSize     = wrw_invoke_params.workSpaceSize;
+            float zero                    = 0.f;
 
-            // Only fp32 use atomic_add, which accumulate the result into dw.
-            // Therefore the need for setting to zero beforehand.
-            if(tensors.dwDesc.GetType() == miopenFloat)
+            if(workspaceSize > 0)
```

Comment:
It's sad to say, but unfortunately this is incorrect.

`workspaceSize` here is a size of workspace actually passed from the user's application. It may be smaller than required (so we must check and issue en error if so) or it can be bigger then required (and there should be no harm in this case).

Let's consider the latter. If kernel does not require workspace, but a user's app passes non-zero workspace, then this branch would take wrong path.

- Please use required workspace size here.
- Please also check if `workspaceSize` >= required workspace.

Reply:
@atamazov Please let me know if my understanding is correct. I'm not that familiar with the interaction between invoker factor, solver, and solution:

- change `MakeMlirWrWInvokerFactor` which takes another extra parameter about the required workspace size, which is returned by MLIR kernel generator.
- change `MakeMlirConvArgs` to use the required workspace size, instead of `bool`.
- inside the double lambda returned by the invoker factory, apply the two checks you mentioned.

Reply:
@atamazov Please help review the latest commit. I tried to address all the feedback based on my understanding.

Reply:
> @atamazov Please let me know if my understanding is correct.



> 
>     * change `MakeMlirWrWInvokerFactor` which takes another extra parameter about the required workspace size, which is returned by MLIR kernel generator.

Yes. Small clarification: required workspace size is returned by solver's `GetWorkspaceSize()` (maybe it asks MLIR kernel gen for it, but this is implementation detail).

>     * change `MakeMlirConvArgs` to use the required workspace size, instead of `bool`.

This is just a recommendation. I hope that there is no cases when GetWorkspaceSize() returns 0 but MLIR WrW kernel still need zero-sized workspace to be passed in.

>     * inside the double lambda returned by the invoker factory, apply the two checks you mentioned.

[Informative] Outer lambda is Invoker Factory. Inner one is Invoker.

In the Invoker, pass workspace-related arguments to the kernel when required workspace size is > 0.

Also check if passed workspace is of enough size.

```C++
if(invoke_params.workSpaceSize < required workspace size)
    MIOPEN_THROW("Not enough workspace for ...");
```

Reply:
@atamazov Got it. I think I addressed your feedback in the latest patch.